cadcutils/cadcutils/net/ws.py

```diff
@@ -461,36 +480,176 @@ def upload_file(self, url, src, md5_checksum=None, **kwargs):
+        seg_size = self._get_segment_size(stat_info.st_size,
+                                          min_segment,
+                                          max_segment)
+        current_size = 0
+        last_digest = hashlib.md5()
+        try:
+            # Obs -(-stat_info.st_size//seg_size) - ceiling division in PYTHON
+            for segment in range(0, -(-stat_info.st_size//seg_size)):
+                cur_seg_size = min(seg_size,
+                                   stat_info.st_size-segment*seg_size)
+                self.logger.debug('Seding segment {} of size {}'.format(
+                    segment, cur_seg_size))
+                # Note: setting the content length here is irrelevant as
+                # requests is going to override it according to the size
+                # of the data (as returned by Md5File file handler)
+                kwargs['headers'] = {PUT_TXN_ID: trans_id,
+                                     HTTP_LENGTH: str(cur_seg_size)}
+                current_size += cur_seg_size
+                retries = 3
+                while retries:
+                    try:
+                        with util.Md5File(src, 'rb', segment*seg_size,
+                                          cur_seg_size) as reader:
+                            reader._md5_checksum = last_digest.copy()
+                            response = self._get_session().put(
+                                url,
+                                data=reader,
+                                verify=self.verify,
+                                **kwargs)
+                    except exceptions.TransferException as e:
+                        self.logger.warning('Errors transfering {} to {}: {}'.
+                                            format(src, url, e.msg))
+                        try:
+                            response = self._get_session().head(
+                                url,
+                                headers={PUT_TXN_ID: trans_id,
+                                         HTTP_LENGTH: str(current_size)})
+                        except Exception as e:
+                            self.logger.error(
+                                'BUG: could not read transaction {} '
+                                'status from {}: {}'.format(trans_id, url,
+                                                            str(e)))
+                            raise e
+                    # check the file made it OK
+                    src_md5 = reader.md5_checksum
+                    dest_md5 = net.extract_md5(response.headers)
+                    if src_md5 != dest_md5:
+                        msg = 'File {} not properly uploaded. ' \
+                              'Mismatched md5 src vs dest: {} vs {}'.format(
+                                src, src_md5, dest_md5)
+                        self.logger.warning(msg)
+                        retries -= 1
+                        if retries:
```

Comment:
I could be mis-reading the code, but if there are no retries left, does cleanup from a botched segment get skipped? And the next segment is sent regardless?

Reply:
When no retries are left, the exception is thrown causing the entire transaction to be aborted no just a segment.